woocommerce/woocommerce
# Support fetching order types in bulk.

### All Submissions:






### Changes proposed in this Pull Request:



Addresses a performance regression where n+1 query pattern was found while loading orders. This n+1 queries were happening because on reading each order, we were doing a DB call to get its order type.

This PR fixes the issue by fetching the order types in bulk.

(Note that we have currently two of these query antipatterns, the second one is for metadata which will be fixed in a separate PR).

### How to test the changes in this Pull Request:

Functionality is already covered by unit test `test_get_orders_with_multiple_order_type`. 

To test performance regression, use the following snippet (after switching to HPOS as data store):
```php
define( 'SAVEQUERIES', true ); // No need for this if it's already defined to true.
global $wpdb;
$query_executed = get_num_queries();
$orders = wc_get_orders(array('limit'=>1000));
foreach( $orders as $order ) {
    assert( $order->get_id() > 0 );
}
$final_query_count = get_num_queries() - $query_executed;
echo $final_query_count; // On trunk -> 2x,  On this branch -> 1x (After fix for meta n+1, this will reduce to about 2-3 irrespective of number of orders fetched)
```

You should see a considerable reduction in query count when on this branch.


### Other information:




### FOR PR REVIEWER ONLY:
